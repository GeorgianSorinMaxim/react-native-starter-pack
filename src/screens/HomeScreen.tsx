import React, { useRef, useEffect, useCallback } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  AppState,
  FlatList,
  StyleSheet,
  View,
  AppStateStatus,
} from "react-native";

import Colors from "../constants/Colors";
import { BodyText, Card } from "../components";

import { getRestaurants } from "../store/selectors";
import { University } from "../store/reducers/data";

import { appActions } from "../store/actions/app";
import { authActions } from "../store/actions/auth";

import { StringValues } from "../constants/StringValues";

export const HomeScreen = () => {
  const dispatch = useDispatch();

  const restaurants = useSelector(getRestaurants);

  const appState = useRef(AppState.currentState);

  const handleAppStateChange = useCallback((nextAppState: AppStateStatus) => {
    if (
      appState.current.match(/inactive|background/) &&
      nextAppState === "active"
    ) {
      dispatch(authActions.validateTokenStart());
    }

    dispatch(
      appActions.appStateUpdated({
        prevState: appState.current,
        newState: nextAppState,
      }),
    );
    appState.current = nextAppState;
  }, []);

  useEffect(() => {
    const subscription = AppState.addEventListener("change", nextAppState =>
      handleAppStateChange(nextAppState),
    );

    return () => {
      subscription.remove();
    };
  }, [handleAppStateChange]);

  const renderItem = (item: University) => {
    return <Card item={item} />;
  };

  return (
    <View style={styles.screenContainer}>
      {restaurants.length > 0 ? (
        <View style={styles.listContainer}>
          <FlatList
            data={restaurants}
            keyExtractor={({ url }) => url}
            renderItem={({ item }) => renderItem(item)}
            style={styles.listContainer}
            showsVerticalScrollIndicator={false}
          />
        </View>
      ) : (
        <BodyText style={styles.centeredText}>
          {StringValues.noDataAvailable}
        </BodyText>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  screenContainer: {
    flex: 1,
    paddingVertical: 12,
    backgroundColor: Colors.white,
  },
  listContainer: {
    marginHorizontal: 12,
  },
  centeredText: {
    fontSize: 16,
    textAlign: "center",
  },
});
