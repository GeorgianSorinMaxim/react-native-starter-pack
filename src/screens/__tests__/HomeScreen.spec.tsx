import * as React from "react";
import renderer from "react-test-renderer";

import { HomeScreenBase } from "../HomeScreen";

jest.mock("@react-navigation/stack");
jest.mock("@react-navigation/native");
jest.mock("@react-navigation/bottom-tabs");

describe("HomeScreen", () => {
  it(`renders the HomeScreen screen`, () => {
    const defaultProps = {
      restaurants: [],
    };

    const tree = renderer.create(<HomeScreenBase {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
