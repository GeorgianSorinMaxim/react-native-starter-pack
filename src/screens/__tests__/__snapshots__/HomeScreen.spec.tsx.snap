// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders the HomeScreen screen 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFF",
      "flex": 1,
      "paddingVertical": 12,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 16,
          "textAlign": "center",
        },
      ]
    }
  >
    No data available
  </Text>
</View>
`;
